proposals/module-types/Explainer.md

```diff
@@ -0,0 +1,725 @@
+shared-everything dynamic linking is controlling exactly which module instances
+share memory. A design where sharing code equates to sharing memory may seem
+natural, but this sacrifices isolation. Lack of isolation can lead to difficult
+bugs that only appear when independent programs are composed. Lack of isolation
+also enables new kinds of supply-chain attacks. Ideally, to maximize isolation,
+each "program" (for some definition of "program" defined by the toolchain)
+should be able to generate a fresh memory each time the program is run,
+regardless of which of its modules are shared with other programs.
+
+For example, it should be possible to take the static dependency graph of the
+application on the left, which contains two programs (`zipper` and `imgmgk`) and
+three shared modules (`libc`, `libzip` and `libimg`), and create the
+dynamically-linked instance graph on the right at application runtime.
+
+<p align="center"><img src="./shared-everything-dynamic-linking.svg" width="600"></p>
+
+Moreover, the linking mechanism must allow the developers of each individual
+module, program and application to independently deal with the inevitable
+version changes (major and minor) of their dependencies.
+
+
+## Additional Requirements
+
+In addition to being able to satisfy the above use cases, there are two more
+general requirements to be considered:
+
+* **No GC dependency**: Since we wish to use this feature to implement dynamic
+  linking of C/C++/Rust programs, and since C/C++/Rust programs aim to run on
+  hosts that don't contain a GC, this proposal should not require a general
+  garbage collection algorithm to implement; reference-counting should suffice.
+
+* **Enable AOT compilation**: Today, using static linking of libraries, wasm
+  hosts are able to use a simple, Ahead-of-Time compilation model to generate
+  high-performance machine code images that can be immediately instantiated and
+  run. Switching to dynamic linking should not hurt hosts' ability to perform
+  the same degree of optimization.
+
+This proposal is checked against these requirements [below](#additional-requirements-revisited).
+
+
+## Proposed Additions
+
+The proposal introduces a set of additions to WebAssembly centered around giving
+modules the ability to import modules. Thus, while the proposal is named "Module
+Imports", there are actually several other additions. The complete list is
+summarized [at the end](#summary).
+
+
+### Single-level Imports
+
+An additional form of import is added which has only one import string instead
+of the usual two:
+```wasm
+(module
+  (import "a" "b" (func))  ;; legal today
+  (import "x" (func))      ;; illegal today, legalized with this proposal
+)
+```
+
+In general, this addition removes a long-standing source of awkwardness on the
+Web: when [ESM-integration] isn't used, the first ("module") name string is
+usually unnecessary leading tools to unnecessarily invent ad hoc module names
+like `env`.
```

Comment:
In the non-ESM (so: JS API) case, the strings are just plain keys that are matched by `instantiate` et al in the same way as they are today... just in a single-level, instead of two-level hierarchy.  I think both cases stay symmetric to the two-level import names, yes?